lib/SPIRV/libSPIRV/SPIRVInstruction.h

```diff
@@ -2721,61 +2721,67 @@ _SPIRV_OP(PowN, true, 10)
-      else
-        llvm_unreachable(
-            "AtomicFAddEXT can only be generated for f32 or f64 types");
-    }
-    // Per the spec OpAtomicCompareExchangeWeak, OpAtomicFlagTestAndSet and
-    // OpAtomicFlagClear instructions require kernel capability. But this
-    // capability should be added by setting OpenCL memory model.
-    return CapVec;
+    // In SPIRV 1.2 spec, only 2 atomic instructions have no result type:
+    // 1. OpAtomicStore - need to check type of the Value operand
+    // 2. OpAtomicFlagClear - doesn't require Int64Atomics capability.
+    // Besides, OpAtomicCompareExchangeWeak, OpAtomicFlagTestAndSet and
+    // OpAtomicFlagClear instructions require the "kernel" capability. But this
+    // capability should be added by setting the OpenCL memory model.
+    if (hasType() && getType()->isTypeInt(64))
+      return {CapabilityInt64Atomics};
+    return {};
   }
 
   // Overriding the following method because of particular OpAtomic*
-  // instructions. We may have to declare additional capabilities,
-  // e.g. based on operand types.
+  // instructions that declare additional capabilities, e.g. based on operand
+  // types.
+  void setOpWords(const std::vector<SPIRVWord> &TheOps) override {
+    SPIRVInstTemplateBase::setOpWords(TheOps);
+    for (auto RC : getRequiredCapability())
+      Module->addCapability(RC);
+  }
+};
+
+class SPIRVAtomicStoreInstBase : public SPIRVAtomicInstBase {
+public:
+  // Overriding the following method because of 'const'-related
+  // issues with overriding getRequiredCapability(). TODO: Resolve.
   void setOpWords(const std::vector<SPIRVWord> &TheOps) override {
     SPIRVInstTemplateBase::setOpWords(TheOps);
     static const unsigned ValueOperandIndex = 3;
-    if (getOpCode() == OpAtomicStore &&
-        getOperand(ValueOperandIndex)->getType()->isTypeInt(64))
+    if (getOperand(ValueOperandIndex)->getType()->isTypeInt(64))
       Module->addCapability(CapabilityInt64Atomics);
-    if (getOpCode() == OpAtomicFAddEXT)
-      for (auto RC : getRequiredCapability())
-        Module->addCapability(RC);
+  }
+};
+
+class SPIRVAtomicFAddEXTInstBase : public SPIRVAtomicInstBase {
+public:
+  llvm::Optional<ExtensionID> getRequiredExtension() const override {
+    return ExtensionID::SPV_EXT_shader_atomic_float_add;
+  }
+
+  SPIRVCapVec getRequiredCapability() const override {
+    assert(hasType());
+    if (getType()->isTypeFloat(32))
+      return {CapabilityAtomicFloat32AddEXT};
+    assert(getType()->isTypeFloat(64) &&
+           "AtomicFAddEXT can only be generated for f32 or f64 types");
+    return {CapabilityAtomicFloat64AddEXT};
   }
 };
 
 #define _SPIRV_OP(x, ...)                                                      \
   typedef SPIRVInstTemplate<SPIRVAtomicInstBase, Op##x, __VA_ARGS__> SPIRV##x;
+#define _SPIRV_OP_SPECIAL(x, ...)                                              \
+  typedef SPIRVInstTemplate<SPIRV##x##InstBase, Op##x, __VA_ARGS__> SPIRV##x;
```

Comment:
I might have to rework this into something like:
```suggestion
#define _SPIRV_OP_SPECIAL(x, BaseName, ...)                                              \
  typedef SPIRVInstTemplate<SPIRV##BaseName##InstBase, Op##x, __VA_ARGS__> SPIRV##x;
```
to allow for more versatile specialization.